File: sources/Olympiad in Informatics/USACO/2011 US Open/G1.cpp
/*
 * USA Computing Olympiad, USACO US Open 2011 Contest, Gold 1. Mowing the Lawn
 * JUNGOL 3621
 *
 * DP, sliding window min/max with deque
 * Date: 2026.5.1
 */

#include<bits/stdc++.h>
using namespace std;
using i64 = int64_t;

int main() {
    cin.tie(0)->sync_with_stdio(0);

    int n, k; cin >> n >> k;
    i64 psum = 0;
    i64 ans = 0;

    deque<pair<int64_t, int> > dq;
    dq.emplace_back(0, -1);
    for (int i = 0; i < n; ++i) {
        int e; cin >> e;
        psum += e;
        i64 x = ans - psum;
        while (!dq.empty() && dq.back().first <= x) dq.pop_back();
        if (!dq.empty() && dq.front().second + k < i) dq.pop_front();
        ans = max(ans, psum + dq.front().first);
        dq.emplace_back(x, i);
    }

    cout << ans;
    return 0;
}

File: sources/Olympiad in Informatics/USACO/2008 Feb/G1.py
"""
# USA Computing Olympiad, USACO February 2008 Contest, Gold 1. Making the Grade
# JUNGOL 4980
#
# greedy, DP
# Date: 2026.5.6
"""

def main() -> None:
    import sys
    rd = lambda: sys.stdin.readline().rstrip()

    n = int(rd())
    a = [int(rd()) for _ in range(n)]
    sa = sorted(a)

    f = [[0] * n for _ in range(n)]

    for j in range(n):
        f[0][j] = abs(a[0] - sa[j])

    for i in range(1, n):
        g = 2**60
        for j in range(n):
            g = min(g, f[i - 1][j])
            f[i][j] = g + abs(a[i] - sa[j])

    print(min(f[-1]))


if __name__ == "__main__":
    main()
